[FINNA-3167] IIIF manifest generator

ronja-koistinen opened the pull request:
Add a IIIFManifest action in RecordController. It generates a IIIF Presentation API v3 manifest for internal use. This is not intended to be exposed in the Finna API or as a user-discoverable resource.

* Pixel size is unavailable in metadata, so it is omitted, even though the Presentation API spec requires it.
* Because images are served over Cover/Show, they are all already image/jpeg, so we hard-code the MIME type
* Configurable CORS headers

See https://iiif.io/api/presentation/3.0/

ronja-koistinen committed:
[FINNA-3167] IIIF manifest generator

Add a IIIFManifest action in RecordController. It generates a IIIF Presentation
API v3 manifest for internal use. This is not intended to be exposed in the
Finna API or as a user-discoverable resource.

* Pixel size is unavailable in metadata, so it is omitted, even though the Presentation API spec requires it.
* Because images are served over Cover/Show, they are all already image/jpeg, so we hard-code the MIME type
* Configurable CORS headers with '*' default

See https://iiif.io/api/presentation/3.0/

ronja-koistinen committed:
Linting

ronja-koistinen commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -882,4 +882,37 @@ protected function showTab($tab, $ajax = false)
+        }
+        if ($manifest = $generator->generate($driver)) {
+            if ($manifestJson = json_encode($manifest)) {
+                $headers->addHeaderLine('Content-Type: application/json');
```
Pitäisikö olla `'Content-Type: application/json;charset=utf-8'`?

EreMaijala replied:
Ei, UTF-8 on oletus JSON:lle.

ronja-koistinen commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,181 @@
+        } else {
+            return $manifest;
+        }
+    }
```
Vaiko näin:

```php
return empty($manifest['items']) ? null : $manifest
```

EreMaijala replied:
Kun `$manifest['items']` on aina array, niin tämä on yksinkertaisin:
```suggestion
        return $manifest['items'] ? $manifest : null;
    }
```

EreMaijala replied:
Toinen vaihtoehto on koota `$items`-array ennen kuin kasaat manifestin. Jos ei löydy mitään, niin voit palauttaa null:n jo ennen kuin haet edes $recordTitle:a tai alat kasaamaan `$manifest`:ia.

ronja-koistinen committed:
Use recommended class references

ronja-koistinen committed:
Improvements

ronja-koistinen committed:
Fixes + rights link

ronja-koistinen commented:
Entä pitäisikö olla `requiredStatement`, jossa mainittaisiin jotain että "Tämä manifesti on Finnan automaattisesti luoma, eikä se välttämättä sovellu monipuoliseen IIIF-jatkokäyttöön"? https://iiif.io/api/presentation/3.0/#requiredstatement

ronja-koistinen committed:
JSON profile

ronja-koistinen committed:
Add more metadata

EreMaijala commented on local/config/finna/config.ini.sample:
```diff
@@ -1963,3 +1963,8 @@ site_status = production
+; Origins to be listed in Access-Control-Allow-Origin headers in the
+; RecordController IIIFManifest action. Recommended to allow all origins for
+; general IIIF use
+recordControllerManifestCORS[] = "*"
```
Laittaisin nyt pois päältä tuon oletuksena (ja kuvauksen sen mukaisesti) niin kauan kun ei haluta, että tuota käytetään Finnan ulkopuolelta. Lisäksi jättäisin `recordController`-alun pois, ettei tämä vanhene käsiin refaktoroinneissa.
```suggestion
;manifestCORS[] = "*"
```

EreMaijala commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -882,4 +882,40 @@ protected function showTab($tab, $ajax = false)
+            \Finna\Record\IIIF\IIIFManifestGenerator::class
+        );
+        $config = $this->getConfigArray();
+        $corsAllow = $config['IIIF']['recordControllerManifestCORS'] ?? ['*'];
```
...ja oletuksena ei olisi CORS sallittu.

EreMaijala commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -882,4 +882,40 @@ protected function showTab($tab, $ajax = false)
+            } else {
+                $headers->addHeaderLine('Content-Type: text/plain');
+                $response->setStatusCode(500);
+                $response->setContent('Error encoding JSON');
```
Tätä ei tietenkään pitäisi tapahtua, mutta jos kuitenkin tapahtuu, niin olisi kiva laittaa virhelokiin ainakin tietueen id ja `json_last_error_msg()`.

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,234 @@
+        return ($this->url)(
+            'cover-show',
+            [],
+            ['force_canonical' => true]
```
Laita query-parametrit [url-helperin kautta](https://docs.laminas.dev/laminas-view/v2/helpers/url/#query-string-arguments).

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Builds the cover URL for this image
```
```suggestion
     * Build the cover URL for this image
```

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Creates annotation page representing a given image
```
```suggestion
     * Create annotation page representing a given image
```

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @return string
+     */
+    private function createBodyId(string $recordId, int $index, string $size, string $source)
```
private:a ei käytetä, kaikki on protected. Lisää paluuarvon tyyppi.

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGeneratorFactory.php:
```diff
@@ -0,0 +1,73 @@
+        $viewRenderer = $container->get('ViewRenderer');
+        $generator = new IIIFManifestGenerator(
+            $viewRenderer->plugin('Url'),
+            $viewRenderer->plugin('ServerUrl'),
```
Ei sillä sinänsä väliä, mutta vakiintunut käytäntö on ollut kirjoittaa pluginien nimet pienellä alkukirjaimella (ehkä siksi, että niitä käytetään samaan tapaan templaateissakin).

ronja-koistinen committed:
Apply suggestions from code review

Co-authored-by: Ere Maijala <ere.maijala@helsinki.fi>

ronja-koistinen committed:
Changes from code review + other slight refactoring

ronja-koistinen commented:
Mitäs me tehdään sen redirectin kanssa, jota ehdotin Slackissä? Luonnostelin jotain tällaista:

```php
    protected function iiifManifestAction()
    {
        $response = $this->getResponse();
        $headers = $response->getHeaders();
        $driver = $this->loadRecord();

        if($datasourceManifests = $driver->tryMethod('getIiifManifests')) {
            $headers->addHeaderLine('Location', $datasourceManifests[0]['url']);
            $response->setStatusCode(301);
            return $response;
        }
    /* jne. */
```

edit: Varmaan 302 parempi itse asiassa.

ronja-koistinen committed:
Add charset in content-type

ronja-koistinen committed:
Redirect to existing manifest + method to get generated manifest URL

ronja-koistinen committed:
Linting fixes

EreMaijala commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -882,4 +882,53 @@ protected function showTab($tab, $ajax = false)
+                $headers->addHeaderLine('Content-Type: text/plain');
+                $response->setStatusCode(500);
+                $response->setContent('Error encoding JSON');
+                error_log(
```
Käytä VuFindin loggeria. Sille on LoggerAwareInterface ja LoggerAwareTrait, joilla sen saa helposti käyttöön.

EreMaijala commented on module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php:
```diff
@@ -0,0 +1,251 @@
+            $manifestItems[] = $canvasItem;
+        }
+
+        if (empty($manifestItems)) {
```
```suggestion
        if (!$manifestItems) {
```

ronja-koistinen committed:
Apply changes from review

EreMaijala commented:
> Mitäs me tehdään sen redirectin kanssa, jota ehdotin Slackissä? Luonnostelin jotain tällaista:
> 
> ```
>     protected function iiifManifestAction()
>     {
>         $response = $this->getResponse();
>         $headers = $response->getHeaders();
>         $driver = $this->loadRecord();
> 
>         if($datasourceManifests = $driver->tryMethod('getIiifManifests')) {
>             $headers->addHeaderLine('Location', $datasourceManifests[0]['url']);
>             $response->setStatusCode(301);
>             return $response;
>         }
>     /* jne. */
> ```
> 
> edit: Varmaan 302 parempi itse asiassa.

Joo, 302 on parempi. Ja toteutus on ok.

ronja-koistinen committed:
Split manifest generation to one more function

This is in anticipation of later unit testing. Now createManifest() does not
require a RecordDriver as input, so it will require no complicated mocking to
unit test.

EreMaijala commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -54,6 +54,7 @@ class RecordController extends \VuFind\Controller\RecordController
     use FinnaRecordControllerTrait;
     use \Finna\Statistics\ReporterTrait;
     use FinnaRecordPreviewSupportTrait;
+    use \VuFind\Log\LoggerAwareTrait;
```
Luokan pitää vielä toteuttaa LoggerAwareInterface, muuten loggeria ei aseteta.

EreMaijala commented on module/Finna/src/Finna/Controller/RecordController.php:
```diff
@@ -882,4 +883,52 @@ protected function showTab($tab, $ajax = false)
+
+        $generator = $this->serviceLocator->get(
+            \Finna\Record\IIIF\IIIFManifestGenerator::class
+        );
```
```suggestion
        $generator = $this->serviceLocator->get(\Finna\Record\IIIF\IIIFManifestGenerator::class);
```

ronja-koistinen committed:
Changes from code review
